client/src/controllers/PreviewController.ts

```diff
@@ -0,0 +1,789 @@
+          // (hence available), we want to clear it immediately to show the
+          // "Preview is not available" screen instead of the outdated preview.
+          if (data.is_available) {
+            this.updatePromise = this.clearPreviewData().then(() => false);
+          } else {
+            // There is no stale data, but we still need to load the iframe to
+            // show the "Preview is not available" screen, because initially
+            // the iframe is empty (to prevent loading the iframe when the panel
+            // is never opened).
+            this.reloadIframe();
+          }
+        } else {
+          // Finish the process when the data is invalid to prepare for the next update
+          // and avoid elements like the loading spinner to be shown indefinitely
+          this.finishUpdate();
+        }
+
+        return data.is_valid as boolean;
+      } catch (error) {
+        this.dispatch('error', { cancelable: false, detail: { error } });
+        this.finishUpdate();
+        // Re-throw error so it can be handled by setPreviewDataWithAlert
+        throw error;
+      }
+    })();
+
+    return this.updatePromise;
+  }
+
+  /**
+   * Clears the preview data from the session.
+   * @returns `Response` from the fetch `DELETE` request
+   */
+  async clearPreviewData() {
+    return fetch(this.urlValue, {
+      headers: {
+        [WAGTAIL_CONFIG.CSRF_HEADER_NAME]: WAGTAIL_CONFIG.CSRF_TOKEN,
+      },
+      method: 'DELETE',
+    }).then((response) => {
+      this.available = false;
+      this.reloadIframe();
+      return response;
+    });
+  }
+
+  /**
+   * Reloads the preview iframe.
+   *
+   * Instead of reloading the iframe with `iframe.contentWindow.location.reload()`
+   * or updating the `src` attribute, this works by creating a new iframe that
+   * replaces the old one once the new one has been loaded. This prevents the
+   * iframe from flashing when reloading.
+   */
+  reloadIframe() {
+    const loadEvent = this.dispatch('load');
+    if (loadEvent.defaultPrevented) {
+      // The load event is cancelled, so don't reload the iframe
+      // and immediately finish the update
+      this.finishUpdate();
+      return;
+    }
+
+    // Copy the iframe element
+    const newIframe = this.iframeTarget.cloneNode() as HTMLIFrameElement;
+
+    // Remove the ID to avoid duplicate IDs in the DOM
+    newIframe.removeAttribute('id');
+
+    // The iframe does not have an src attribute on initial load,
+    // so we need to set it here. For subsequent loads, it's fine to set it
+    // again to ensure it's in sync with the selected preview mode.
+    const url = this.renderUrl;
+    url.searchParams.set('in_preview_panel', 'true');
+    newIframe.src = url.toString();
+
+    // Make the new iframe invisible
+    newIframe.style.width = '0';
+    newIframe.style.height = '0';
+    newIframe.style.opacity = '0';
+    newIframe.style.position = 'absolute';
```

Comment:
Is it worth us pulling this into classes?

We already have this in the preview scss file.

```
    input[type='radio'] {
      position: absolute;
      width: 0;
      height: 0;
      opacity: 0;
    }
```

Maybe we could just add a new class here such as `w-preview-proxy`?